In [19]:
from sklearn.decomposition import PCA
from skimage import io, color,transform
from skimage.metrics import structural_similarity as ssim
import numpy as np

# Load the images
image1 = io.imread('OIP1.jpg')
image2 = io.imread('OIP2.jpg')

# Convert the images to grayscale
gray_image1 = color.rgb2gray(image1)
gray_image2 = color.rgb2gray(image2)

# Resize the images to a common size
resized_image1 = transform.resize(gray_image1, (256, 256))
resized_image2 = transform.resize(gray_image2, (256, 256))

# Extract features from the two grayscale images using supervised approach
feature_vector1 = resized_image1.flatten()
feature_vector2 = resized_image2.flatten()

# Normalize the feature vectors
feature_vector1 = feature_vector1 / np.linalg.norm(feature_vector1)
feature_vector2 = feature_vector2 / np.linalg.norm(feature_vector2)

# Compute the cosine similarity between the two feature vectors
cosine_similarity_supervised = np.dot(feature_vector1, feature_vector2)

# Extract features from the two grayscale images using unsupervised approach
pca = PCA(n_components=1).fit(resized_image1)
feature_vector1_unsupervised = pca.transform(resized_image1).ravel()

pca = PCA(n_components=1).fit(resized_image2)
feature_vector2_unsupervised = pca.transform(resized_image2).ravel()

# Normalize the feature vectors
feature_vector1_unsupervised = feature_vector1_unsupervised / np.linalg.norm(feature_vector1_unsupervised)
feature_vector2_unsupervised = feature_vector2_unsupervised / np.linalg.norm(feature_vector2_unsupervised)

# Compute the cosine similarity between the two feature vectors
cosine_similarity_unsupervised = np.dot(feature_vector1_unsupervised, feature_vector2_unsupervised)

# Print the similarity scores
print("Supervised Cosine Similarity:", cosine_similarity_supervised)
print("Unsupervised Cosine Similarity:", cosine_similarity_unsupervised) 

Supervised Cosine Similarity: 0.934118317439667
Unsupervised Cosine Similarity: 0.8510803699479844


Based on the code, we can see that the images are first loaded and converted to grayscale. 
Then, both supervised and unsupervised approaches are used to extract features from the images. 
The feature vectors are then normalized and cosine similarity is computed for both approaches.

The supervised approach uses the flattened resized images as feature vectors, whereas the unsupervised approach 
uses PCA to extract a single principal component from the resized images as feature vectors.

Comparing the two similarity scores, we can see that the supervised approach yields a higher similarity 
score of 0.9341, whereas the unsupervised approach yields a lower similarity score of 0.8511.

Therefore, based on this analysis, the supervised approach appears to be better at capturing the similarity 
between the two images. However, its worth noting that the choice of approach may depend on the specific
use case and the type of images being analyzed.